### PR TITLE
quality range not checked leading to undefined behavior

### PR DESCRIPTION
**Pull request by joe-angell**
Quality scores according to the fastq wiki can be up to 93, so updating stats to allow that.  Also check to make sure we aren't writing out of bounds.

Some of the stats classes were missing virtual destructors which also leads to undefined behavior.

### CODE REVIEW

**Comment by joe-angell** [comment]
@msettles @samhunter @dstreett someone give a quick review please, i'm not sure how you want to handle quality scores outside the range you had set.  According to wikipedia "Sanger format can encode a [Phred quality score](https://en.wikipedia.org/wiki/Phred_quality_score) from 0 to 93 using ASCII 33 to 126" so I raised it for stats.

**Comment by samhunter** [comment]
I think the correct fix for this is to add a "-q64" (or equivalent) switch to support this old Illumina format:

> "Starting with Illumina 1.3 and before Illumina 1.8, the format encoded a [Phred quality score](https://en.wikipedia.org/wiki/Phred_quality_score) from 0 to 62 using [ASCII](https://en.wikipedia.org/wiki/ASCII) 64 to 126 (although in raw read data Phred scores from 0 to 40 only are expected)."
> --https://en.wikipedia.org/wiki/FASTQ_format#Encoding


@msettles with various new data types becoming available and potentially larger range of valid quality values, I wonder if the 0-40 range should be expanded to support 0-93 according to the Sanger format as @joe-angell rightly points out above. What do you think? 

From my side, I don't see any problems with this, except that currently when we overlap PE reads we cap the q-score for matches at 40 to fit with Illumina encoding. I think we could probably still do this since the q-score is an approximation at that point anyway.

File: hts_AdapterTrimmer/src/hts_AdapterTrimmer.h [hunk 1]
```diff
@@ -137,12 +138,12 @@ class AdapterTrimmer: public MainTemplate<AdapterCounters, AdapterTrimmer> {
                 read1_bp = loc1_t + i;
                 read2_bp = loc2_t + i;
                 if (seq1[read1_bp] == seq2[read2_bp] )  {
-                    qual = static_cast<char>(std::min(qual1[read1_bp] + qual2[read2_bp] - 33, 40 + 33));  // MATT: I still don't agree with this, so 38 [confident] and 2 [not so confident] return a 40 [very confident]??
+                    qual = static_cast<char>(std::min(qual1[read1_bp] + qual2[read2_bp] - qual_offset, 40 + qual_offset));  // MATT: I still don't agree with this, so 38 [confident] and 2 [not so confident] return a 40 [very confident]??
```
Comment (joe-angell) [inline hts_AdapterTrimmer/src/hts_AdapterTrimmer.h:141]: Do we need a max quality option here for adapter trimmer?

Comment (samhunter) [reply hts_AdapterTrimmer/src/hts_AdapterTrimmer.h:141]: @joe-angell sorry for the slow reply. 

I think it is still appropriate to set a maximum. My argument being that adapter trimmer and overlapper in particular were designed specifically to support Illumina libraries and the Illumina paired-read sequencing strategy. 

That being said, according to the Wikipedia article the max should probably be 41 for both the old Phred+64 and new Phred+33:

 _J - Illumina 1.5+ Phred+64,  raw reads typically (3, **41**)
     with 0=unused, 1=unused, 2=Read Segment Quality Control Indicator (bold) 
     (Note: See discussion above).
 L - Illumina 1.8+ Phred+33,  raw reads typically (0, **41**)_

I'm surprised by that, I thought the max was 40, but apparently not!

Also, @msettles can we remove the comment about not agreeing with this approach? This is the definition of a consensus sequence right? Almost exactly what CCS does (e.g. read the same molecule more than once, increase confidence in call if the same thing is seen more than once)?

Comment (joe-angell) [reply hts_AdapterTrimmer/src/hts_AdapterTrimmer.h:141]: @samhunter do you think we can merge this as is (for there are important bug fixes) and address the other issues in another pr?

**Review by dstreett — APPROVED** [review]
I think it looks great @joe-angell  - and if it is causing a crash, we should merge it. (But let's give the other guys 24 hours ;) )

**Comment by samhunter** [comment]
I think we should merge it.

**Pull request merged**